{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "a15dc821-7108-46a9-bad5-4f6f59121f8d",
   "metadata": {},
   "source": [
    "# Workshop 4 - Featurization\n",
    "\n",
    "In this workshop we're going to take the data we reviewed last week during EDA and transform it into usable features to begin modeling. We have a few different options and we'll run through each of the three in turn:\n",
    "1. Gene presence absence\n",
    "2. Kmer counts\n",
    "3. [Bonus] Gene Sequences\n",
    "\n",
    "\n",
    "\n",
    "Lets run through the steps together (there are some questions and some blanks to fill in as we run through)."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d7140deb-bd7d-48db-9a5a-367700da47a3",
   "metadata": {},
   "source": [
    "## Imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "79d595ae-d425-4900-8b62-7136b25e777e",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from collections import defaultdict\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from tensorflow import keras"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e327e0dd-fa16-4a1b-aea9-a2e7b2e17d01",
   "metadata": {},
   "source": [
    "## 1. Load Data\n",
    "\n",
    "Last week we took a look through all our raw data\n",
    "\n",
    "To allow for testing further down the line I've held out some data in advance that we'll be using later in the course. As such I've made a folder in the course project folder system called `train_test_data`\n",
    "\n",
    "For this workshop please download:\n",
    "- `train_test_data` folder and put within `data/`\n",
    "\n",
    "Key for data:\n",
    "- train_genes = gene_match_data for train samples\n",
    "- test_genes = gene_match_data for test samples\n",
    "- y_train = array of S/R target values\n",
    "- y_train_ids = array of genome_ids in order of y_train\n",
    "- y_test_ids = array of genome_ids in order of y_test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "89fd3c81-1867-4819-b907-5f5816a41379",
   "metadata": {},
   "outputs": [],
   "source": [
    "seed = 130\n",
    "\n",
    "def load_data():\n",
    "    \"\"\"\n",
    "    Load the data needed for Workshop 4\n",
    "    \"\"\"\n",
    "    train_genes = pd.read_csv('../data/train_test_data/train_genes.csv')\n",
    "    train_genes['genome_id'] = train_genes.genome_id.astype(str)\n",
    "    test_genes = pd.read_csv('../data/train_test_data/test_genes.csv')\n",
    "    test_genes['genome_id'] = test_genes.genome_id.astype(str)\n",
    "    y_train = np.load('../data/train_test_data/y_train.npy', allow_pickle=True)\n",
    "    y_train_ids = np.load('../data/train_test_data/train_ids.npy', allow_pickle=True).astype(str)\n",
    "    y_test_ids = np.load('../data/train_test_data/test_ids.npy', allow_pickle=True).astype(str)\n",
    "\n",
    "    return train_genes, test_genes, y_train, y_train_ids, y_test_ids\n",
    "\n",
    "train_genes, test_genes, y_train, y_train_ids, y_test_ids = load_data()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "6f73fad7-0542-40c0-b447-e8dc8e6bd7b6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genome_id</th>\n",
       "      <th>contig</th>\n",
       "      <th>res_gene</th>\n",
       "      <th>match_start</th>\n",
       "      <th>match_end</th>\n",
       "      <th>match_qual</th>\n",
       "      <th>query_str</th>\n",
       "      <th>ref_gene_str</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>562.22444</td>\n",
       "      <td>562.22444.con.0012</td>\n",
       "      <td>gb|U00096.3|-|3324062-3324911|ARO:3003386|Ecol...</td>\n",
       "      <td>41274</td>\n",
       "      <td>42123</td>\n",
       "      <td>849M</td>\n",
       "      <td>TTACTCATAGCGTTTGTTTTCCTTTGCAGACAGAGTGGCTTCCACC...</td>\n",
       "      <td>TTACTCATAGCGTTTGTTTTCCTTTGCAGACAGAGTGGCTTCCACC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>562.22444</td>\n",
       "      <td>562.22444.con.0019</td>\n",
       "      <td>gb|AP009048.1|+|3760295-3762710|ARO:3003303|Ec...</td>\n",
       "      <td>60711</td>\n",
       "      <td>63126</td>\n",
       "      <td>2415M</td>\n",
       "      <td>ATGTCGAATTCTTATGACTCCTCCAGTATCAAAGTCCTGAAAGGGC...</td>\n",
       "      <td>ATGTCGAATTCTTATGACTCCTCCAGTATCAAAGTCCTGAAAGGGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>562.22444</td>\n",
       "      <td>562.22444.con.0030</td>\n",
       "      <td>gb|BA000007.3|+|4990267-4994296|ARO:3003288|Ec...</td>\n",
       "      <td>21837</td>\n",
       "      <td>25866</td>\n",
       "      <td>4029M</td>\n",
       "      <td>TTACTCGTCTTCCAGTTCGATGTTGATACCCAGCGAACGAATCTCT...</td>\n",
       "      <td>TTACTCGTCTTCCAGTTCGATGTTGATACCCAGCGAACGAATCTCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>562.22444</td>\n",
       "      <td>562.22444.con.0088</td>\n",
       "      <td>gb|AH002539.2|+|1760-2945|ARO:3003368|Ecol_EFT...</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>3M135I1023M19I5M</td>\n",
       "      <td>GTGTCTAAAGAAAAATTTGAACGTACAAAACCGCACGTTAACGTTG...</td>\n",
       "      <td>GcaTTCGACCAGATCGATAACGCGCCGGAAGAAAAAGCTCGTGGTA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>562.22444</td>\n",
       "      <td>562.22444.con.0088</td>\n",
       "      <td>gb|AE014075.1|-|3901532-3902762|ARO:3003369|Ec...</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>4M180I1022M19I5M</td>\n",
       "      <td>GTGCTCTCTCCTGAAGGGGAGAGCACTATAGTAAGGAATATAGCCG...</td>\n",
       "      <td>GcatTCGACCAGATCGATAACGCGCCGGAAGAAAAAGCTCGTGGTA...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   genome_id              contig  \\\n",
       "0  562.22444  562.22444.con.0012   \n",
       "1  562.22444  562.22444.con.0019   \n",
       "2  562.22444  562.22444.con.0030   \n",
       "3  562.22444  562.22444.con.0088   \n",
       "4  562.22444  562.22444.con.0088   \n",
       "\n",
       "                                            res_gene  match_start  match_end  \\\n",
       "0  gb|U00096.3|-|3324062-3324911|ARO:3003386|Ecol...        41274      42123   \n",
       "1  gb|AP009048.1|+|3760295-3762710|ARO:3003303|Ec...        60711      63126   \n",
       "2  gb|BA000007.3|+|4990267-4994296|ARO:3003288|Ec...        21837      25866   \n",
       "3  gb|AH002539.2|+|1760-2945|ARO:3003368|Ecol_EFT...            0          3   \n",
       "4  gb|AE014075.1|-|3901532-3902762|ARO:3003369|Ec...            0          4   \n",
       "\n",
       "         match_qual                                          query_str  \\\n",
       "0              849M  TTACTCATAGCGTTTGTTTTCCTTTGCAGACAGAGTGGCTTCCACC...   \n",
       "1             2415M  ATGTCGAATTCTTATGACTCCTCCAGTATCAAAGTCCTGAAAGGGC...   \n",
       "2             4029M  TTACTCGTCTTCCAGTTCGATGTTGATACCCAGCGAACGAATCTCT...   \n",
       "3  3M135I1023M19I5M  GTGTCTAAAGAAAAATTTGAACGTACAAAACCGCACGTTAACGTTG...   \n",
       "4  4M180I1022M19I5M  GTGCTCTCTCCTGAAGGGGAGAGCACTATAGTAAGGAATATAGCCG...   \n",
       "\n",
       "                                        ref_gene_str  \n",
       "0  TTACTCATAGCGTTTGTTTTCCTTTGCAGACAGAGTGGCTTCCACC...  \n",
       "1  ATGTCGAATTCTTATGACTCCTCCAGTATCAAAGTCCTGAAAGGGC...  \n",
       "2  TTACTCGTCTTCCAGTTCGATGTTGATACCCAGCGAACGAATCTCT...  \n",
       "3  GcaTTCGACCAGATCGATAACGCGCCGGAAGAAAAAGCTCGTGGTA...  \n",
       "4  GcatTCGACCAGATCGATAACGCGCCGGAAGAAAAAGCTCGTGGTA...  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "train_genes.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "f834e26f-4fb2-4e95-880a-92895f339a60",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(array([['S'],\n",
       "        ['S'],\n",
       "        ['S'],\n",
       "        ['S'],\n",
       "        ['S']], dtype=object),\n",
       " array(['562.22444', '562.42705', '562.42733', '562.42728', '562.45644'],\n",
       "       dtype='<U9'))"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "y_train[0:5], y_train_ids[0:5]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b354afc5-3d28-4f03-9c52-8dea927efed0",
   "metadata": {},
   "source": [
    "## 1. Presence / Absence Features\n",
    "\n",
    "Our first, most simple feature set will be the presence/absence of each gene we've seen from our CARD alignment data.\n",
    "\n",
    "In order to build these features we'll need to:\n",
    "1. Find all unique res_genes, count which samples they're present/absent in\n",
    "2. Look for correlations between genes\n",
    "3. Remove highly correlated features (as seen in the assignment)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0823454f-c4cb-431f-9b2f-4dce1b8299d6",
   "metadata": {},
   "source": [
    "### 1a. Create a presence absence matrix for each sample and each gene\n",
    "\n",
    "- We're going to leverage some pandas magic to make this really simple\n",
    "- The logic can be quite complex manually\n",
    "  - For each gene, search each samples\n",
    "  - Store a list of 0/1 for each samples for each gene\n",
    "  - Ensuring correct ordering\n",
    "- Pandas can do this for us using `pivot_table`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "a6862528-7da5-4af6-967a-ac01113e0a8f",
   "metadata": {},
   "outputs": [],
   "source": [
    "def build_gene_presence_absence(dataset, ids, gene_names=None):\n",
    "    \"\"\"\n",
    "    Build a matrix of samples to genes with 1 for present and 0 for absent\n",
    "\n",
    "    Args:\n",
    "      - dataset (pd.DataFrame): dataset of gene alignments\n",
    "      - ids (list): ordering for IDs\n",
    "      - gene_names (list): ordered list of unique genes (optional)\n",
    "    \"\"\"\n",
    "    # If not providing genes take all unique genes from the data\n",
    "    if gene_names is None:\n",
    "        gene_names = dataset.res_gene.unique()\n",
    "\n",
    "    # Count for all genes found within the data\n",
    "    genes_counts = (\n",
    "        dataset\n",
    "        [['genome_id', 'res_gene']]\n",
    "        .assign(count = 1)\n",
    "        .pivot_table('count', index='genome_id', columns='res_gene')\n",
    "        .fillna(0)\n",
    "    )\n",
    "\n",
    "    # Add genes missing from the data\n",
    "    missing_genes = set(gene_names).difference(set(genes_counts.columns))\n",
    "    if len(missing_genes) > 0:\n",
    "        for gene in missing_genes:\n",
    "            genes_counts[gene] = 0\n",
    "\n",
    "    # Make sure to return in the same order as gene_names and the same sample order\n",
    "    return genes_counts.loc[ids, gene_names]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a49270eb-0613-4a4a-a28f-5e5a1a7f44c0",
   "metadata": {},
   "source": [
    "<div class=\"question\" style=\"color: #534646; background-color: #ffdfa3; padding: 1px; border-radius: 5px;\">\n",
    "\n",
    "#### Q. Why do we need to pass in a list of gene names?\n",
    "\n",
    "</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "963c8ed9-db35-4b98-8921-036dc05253ef",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Shape of PA df: (748, 1428) \n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>res_gene</th>\n",
       "      <th>gb|U00096.3|-|3324062-3324911|ARO:3003386|Ecol_folP_SLF</th>\n",
       "      <th>gb|AP009048.1|+|3760295-3762710|ARO:3003303|Ecol_gyrB_AMU</th>\n",
       "      <th>gb|BA000007.3|+|4990267-4994296|ARO:3003288|Ecol_rpoB_RIF</th>\n",
       "      <th>gb|AH002539.2|+|1760-2945|ARO:3003368|Ecol_EFTu_KIR</th>\n",
       "      <th>gb|AE014075.1|-|3901532-3902762|ARO:3003369|Ecol_EFTu_PLV</th>\n",
       "      <th>gb|U00096.3|-|2336792-2339420|ARO:3003294|Ecol_gyrA_FLO</th>\n",
       "      <th>gb|AP009048.1|-|3172159-3174052|ARO:3003316|Ecol_parE_FLO</th>\n",
       "      <th>gb|U00096.3|-|3470144-3471329|ARO:3003370|Ecol_EFTu_ENC</th>\n",
       "      <th>gb|AE006468.2|-|4038867-4041282|ARO:3003307|Sser_gyrB_FLO</th>\n",
       "      <th>gb|U00096.3|-|3163714-3165973|ARO:3003308|Ecol_parC_FLO</th>\n",
       "      <th>...</th>\n",
       "      <th>gb|NG_067227.1|+|0-900|ARO:3006220|VEB-25</th>\n",
       "      <th>gb|NG_070222.1|+|0-900|ARO:3006221|VEB-26</th>\n",
       "      <th>gb|NG_070755.1|+|0-900|ARO:3006222|VEB-27</th>\n",
       "      <th>gb|KJ158441.1|+|2732-3377|ARO:3002789|QnrD2</th>\n",
       "      <th>gb|FJ228229.1|+|553-1198|ARO:3002788|QnrD1</th>\n",
       "      <th>gb|KF526113.1|+|0-1146|ARO:3002112|CMY-100</th>\n",
       "      <th>gb|AB107899.1|+|1007-2153|ARO:3001856|CFE-1</th>\n",
       "      <th>gb|KF992025.1|+|1026-2172|ARO:3002106|CMY-93</th>\n",
       "      <th>gb|X03988.1|+|382-1642|ARO:3000363|EreB</th>\n",
       "      <th>gb|AJ310778.1|+|7003-7573|ARO:3003015|dfrA19</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>genome_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>562.22444</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>562.42705</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>562.42733</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>562.42728</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>562.45644</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 1428 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "res_gene   gb|U00096.3|-|3324062-3324911|ARO:3003386|Ecol_folP_SLF  \\\n",
       "genome_id                                                            \n",
       "562.22444                                                1.0         \n",
       "562.42705                                                1.0         \n",
       "562.42733                                                1.0         \n",
       "562.42728                                                1.0         \n",
       "562.45644                                                1.0         \n",
       "\n",
       "res_gene   gb|AP009048.1|+|3760295-3762710|ARO:3003303|Ecol_gyrB_AMU  \\\n",
       "genome_id                                                              \n",
       "562.22444                                                1.0           \n",
       "562.42705                                                1.0           \n",
       "562.42733                                                1.0           \n",
       "562.42728                                                1.0           \n",
       "562.45644                                                1.0           \n",
       "\n",
       "res_gene   gb|BA000007.3|+|4990267-4994296|ARO:3003288|Ecol_rpoB_RIF  \\\n",
       "genome_id                                                              \n",
       "562.22444                                                1.0           \n",
       "562.42705                                                1.0           \n",
       "562.42733                                                1.0           \n",
       "562.42728                                                1.0           \n",
       "562.45644                                                1.0           \n",
       "\n",
       "res_gene   gb|AH002539.2|+|1760-2945|ARO:3003368|Ecol_EFTu_KIR  \\\n",
       "genome_id                                                        \n",
       "562.22444                                                1.0     \n",
       "562.42705                                                0.0     \n",
       "562.42733                                                0.0     \n",
       "562.42728                                                1.0     \n",
       "562.45644                                                0.0     \n",
       "\n",
       "res_gene   gb|AE014075.1|-|3901532-3902762|ARO:3003369|Ecol_EFTu_PLV  \\\n",
       "genome_id                                                              \n",
       "562.22444                                                1.0           \n",
       "562.42705                                                0.0           \n",
       "562.42733                                                0.0           \n",
       "562.42728                                                1.0           \n",
       "562.45644                                                0.0           \n",
       "\n",
       "res_gene   gb|U00096.3|-|2336792-2339420|ARO:3003294|Ecol_gyrA_FLO  \\\n",
       "genome_id                                                            \n",
       "562.22444                                                1.0         \n",
       "562.42705                                                1.0         \n",
       "562.42733                                                1.0         \n",
       "562.42728                                                1.0         \n",
       "562.45644                                                1.0         \n",
       "\n",
       "res_gene   gb|AP009048.1|-|3172159-3174052|ARO:3003316|Ecol_parE_FLO  \\\n",
       "genome_id                                                              \n",
       "562.22444                                                1.0           \n",
       "562.42705                                                1.0           \n",
       "562.42733                                                1.0           \n",
       "562.42728                                                1.0           \n",
       "562.45644                                                1.0           \n",
       "\n",
       "res_gene   gb|U00096.3|-|3470144-3471329|ARO:3003370|Ecol_EFTu_ENC  \\\n",
       "genome_id                                                            \n",
       "562.22444                                                1.0         \n",
       "562.42705                                                0.0         \n",
       "562.42733                                                0.0         \n",
       "562.42728                                                1.0         \n",
       "562.45644                                                0.0         \n",
       "\n",
       "res_gene   gb|AE006468.2|-|4038867-4041282|ARO:3003307|Sser_gyrB_FLO  \\\n",
       "genome_id                                                              \n",
       "562.22444                                                1.0           \n",
       "562.42705                                                0.0           \n",
       "562.42733                                                0.0           \n",
       "562.42728                                                1.0           \n",
       "562.45644                                                1.0           \n",
       "\n",
       "res_gene   gb|U00096.3|-|3163714-3165973|ARO:3003308|Ecol_parC_FLO  ...  \\\n",
       "genome_id                                                           ...   \n",
       "562.22444                                                1.0        ...   \n",
       "562.42705                                                1.0        ...   \n",
       "562.42733                                                1.0        ...   \n",
       "562.42728                                                1.0        ...   \n",
       "562.45644                                                1.0        ...   \n",
       "\n",
       "res_gene   gb|NG_067227.1|+|0-900|ARO:3006220|VEB-25  \\\n",
       "genome_id                                              \n",
       "562.22444                                        0.0   \n",
       "562.42705                                        0.0   \n",
       "562.42733                                        0.0   \n",
       "562.42728                                        0.0   \n",
       "562.45644                                        0.0   \n",
       "\n",
       "res_gene   gb|NG_070222.1|+|0-900|ARO:3006221|VEB-26  \\\n",
       "genome_id                                              \n",
       "562.22444                                        0.0   \n",
       "562.42705                                        0.0   \n",
       "562.42733                                        0.0   \n",
       "562.42728                                        0.0   \n",
       "562.45644                                        0.0   \n",
       "\n",
       "res_gene   gb|NG_070755.1|+|0-900|ARO:3006222|VEB-27  \\\n",
       "genome_id                                              \n",
       "562.22444                                        0.0   \n",
       "562.42705                                        0.0   \n",
       "562.42733                                        0.0   \n",
       "562.42728                                        0.0   \n",
       "562.45644                                        0.0   \n",
       "\n",
       "res_gene   gb|KJ158441.1|+|2732-3377|ARO:3002789|QnrD2  \\\n",
       "genome_id                                                \n",
       "562.22444                                          0.0   \n",
       "562.42705                                          0.0   \n",
       "562.42733                                          0.0   \n",
       "562.42728                                          0.0   \n",
       "562.45644                                          0.0   \n",
       "\n",
       "res_gene   gb|FJ228229.1|+|553-1198|ARO:3002788|QnrD1  \\\n",
       "genome_id                                               \n",
       "562.22444                                         0.0   \n",
       "562.42705                                         0.0   \n",
       "562.42733                                         0.0   \n",
       "562.42728                                         0.0   \n",
       "562.45644                                         0.0   \n",
       "\n",
       "res_gene   gb|KF526113.1|+|0-1146|ARO:3002112|CMY-100  \\\n",
       "genome_id                                               \n",
       "562.22444                                         0.0   \n",
       "562.42705                                         0.0   \n",
       "562.42733                                         0.0   \n",
       "562.42728                                         0.0   \n",
       "562.45644                                         0.0   \n",
       "\n",
       "res_gene   gb|AB107899.1|+|1007-2153|ARO:3001856|CFE-1  \\\n",
       "genome_id                                                \n",
       "562.22444                                          0.0   \n",
       "562.42705                                          0.0   \n",
       "562.42733                                          0.0   \n",
       "562.42728                                          0.0   \n",
       "562.45644                                          0.0   \n",
       "\n",
       "res_gene   gb|KF992025.1|+|1026-2172|ARO:3002106|CMY-93  \\\n",
       "genome_id                                                 \n",
       "562.22444                                           0.0   \n",
       "562.42705                                           0.0   \n",
       "562.42733                                           0.0   \n",
       "562.42728                                           0.0   \n",
       "562.45644                                           0.0   \n",
       "\n",
       "res_gene   gb|X03988.1|+|382-1642|ARO:3000363|EreB  \\\n",
       "genome_id                                            \n",
       "562.22444                                      0.0   \n",
       "562.42705                                      0.0   \n",
       "562.42733                                      0.0   \n",
       "562.42728                                      0.0   \n",
       "562.45644                                      0.0   \n",
       "\n",
       "res_gene   gb|AJ310778.1|+|7003-7573|ARO:3003015|dfrA19  \n",
       "genome_id                                                \n",
       "562.22444                                           0.0  \n",
       "562.42705                                           0.0  \n",
       "562.42733                                           0.0  \n",
       "562.42728                                           0.0  \n",
       "562.45644                                           0.0  \n",
       "\n",
       "[5 rows x 1428 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "train_presence_absence_df = build_gene_presence_absence(train_genes, y_train_ids)\n",
    "print('\\nShape of PA df:', train_presence_absence_df.shape, '\\n')\n",
    "train_presence_absence_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "5a4295de-3e15-410d-9a50-692a6a1131d9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>res_gene</th>\n",
       "      <th>gb|AAAGNS010000063.1|-|144-1005|ARO:3000966|TEM-103</th>\n",
       "      <th>gb|AB023477.1|+|0-861|ARO:3001082|SHV-24</th>\n",
       "      <th>Random Missing Gene</th>\n",
       "      <th>gb|AB089595.1|+|0-1206|ARO:3000166|tet(B)</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>genome_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>562.22444</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>562.42705</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>562.42733</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>562.42728</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>562.45644</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "res_gene   gb|AAAGNS010000063.1|-|144-1005|ARO:3000966|TEM-103  \\\n",
       "genome_id                                                        \n",
       "562.22444                                                0.0     \n",
       "562.42705                                                1.0     \n",
       "562.42733                                                1.0     \n",
       "562.42728                                                0.0     \n",
       "562.45644                                                1.0     \n",
       "\n",
       "res_gene   gb|AB023477.1|+|0-861|ARO:3001082|SHV-24  Random Missing Gene  \\\n",
       "genome_id                                                                  \n",
       "562.22444                                       0.0                    0   \n",
       "562.42705                                       0.0                    0   \n",
       "562.42733                                       0.0                    0   \n",
       "562.42728                                       0.0                    0   \n",
       "562.45644                                       0.0                    0   \n",
       "\n",
       "res_gene   gb|AB089595.1|+|0-1206|ARO:3000166|tet(B)  \n",
       "genome_id                                             \n",
       "562.22444                                        0.0  \n",
       "562.42705                                        0.0  \n",
       "562.42733                                        0.0  \n",
       "562.42728                                        1.0  \n",
       "562.45644                                        1.0  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes_of_interest = [\n",
    "    'gb|AAAGNS010000063.1|-|144-1005|ARO:3000966|TEM-103',\n",
    "    'gb|AB023477.1|+|0-861|ARO:3001082|SHV-24',\n",
    "    'Random Missing Gene',\n",
    "    'gb|AB089595.1|+|0-1206|ARO:3000166|tet(B)',\n",
    "]\n",
    "build_gene_presence_absence(train_genes, y_train_ids, genes_of_interest).head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8fa6e620-cf67-4421-bf07-bcf9e3691938",
   "metadata": {},
   "source": [
    "### 2b. Review Correlations and Remove Identical Features\n",
    "\n",
    "As seen in last week's assignment, these presence absence features have a lot of redundancies (genes that are identical across samples).\n",
    "\n",
    "If we leave these, they can cause issues during modeling (unidentifiability/incorrect feature importances and inferences).\n",
    "\n",
    "Two options:\n",
    "1. Calculate correlations and cluster\n",
    "2. Look at identical presence/absence and select one\n",
    "\n",
    "We'll go for the simpler option 2 but you can also try clustering in your project if you believe it will help model performance.\n",
    "\n",
    "Again this seems somewhat complex - we need to look across all samples and check for identical 1/0 arrays. BUT:\n",
    "\n",
    "- Pandas to the rescue once again"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f8b190b0-5781-4a8b-b93c-8dbe1599a7fd",
   "metadata": {},
   "source": [
    "<div class=\"question\" style=\"color: #534646; background-color: #ffdfa3; padding: 1px; border-radius: 5px;\">\n",
    "\n",
    "#### Q. Can anyone think of a common pandas operation which might help use remove identical values?\n",
    "\n",
    "</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "d2ef5f3f-e9c8-40fb-a659-2b3112d6dec1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Transpose the data so that genes are our rows\n",
    "train_presence_absence_transposed = train_presence_absence_df.transpose()\n",
    "\n",
    "# Use \"drop_duplicates()\" to remove identical rows (will just keep the first)\n",
    "train_unique_presence_absence = train_presence_absence_transposed.drop_duplicates()\n",
    "\n",
    "# Transpose back so genes are columns\n",
    "train_presence_absence_df = train_unique_presence_absence.transpose()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "53a564a9-85e5-4cf4-a029-4cde2697f3b7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Shape of PA df: (748, 159) \n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>res_gene</th>\n",
       "      <th>gb|U00096.3|-|3324062-3324911|ARO:3003386|Ecol_folP_SLF</th>\n",
       "      <th>gb|AP009048.1|+|3760295-3762710|ARO:3003303|Ecol_gyrB_AMU</th>\n",
       "      <th>gb|BA000007.3|+|4990267-4994296|ARO:3003288|Ecol_rpoB_RIF</th>\n",
       "      <th>gb|AH002539.2|+|1760-2945|ARO:3003368|Ecol_EFTu_KIR</th>\n",
       "      <th>gb|AE014075.1|-|3901532-3902762|ARO:3003369|Ecol_EFTu_PLV</th>\n",
       "      <th>gb|U00096.3|-|2336792-2339420|ARO:3003294|Ecol_gyrA_FLO</th>\n",
       "      <th>gb|AE006468.2|-|4038867-4041282|ARO:3003307|Sser_gyrB_FLO</th>\n",
       "      <th>gb|HG738867.1|+|2934266-2935658|ARO:3003890|Ecol_UhpT_FOF</th>\n",
       "      <th>gb|HG738867.1|-|2233288-2234647|ARO:3003889|Ecol_GlpT_FOF</th>\n",
       "      <th>gb|HG738867.1|+|2930707-2931298|ARO:3003893|Ecol_uhpA_FOF</th>\n",
       "      <th>...</th>\n",
       "      <th>gb|X01702.1|+|484-1285|ARO:3002659|APH(6)-Ic</th>\n",
       "      <th>gb|JX259319.1|+|0-681|ARO:3002771|QnrB58</th>\n",
       "      <th>gb|EU675686.2|+|3196-4042|ARO:3002616|aadA16</th>\n",
       "      <th>gb|HQ418999.2|+|0-681|ARO:3002746|QnrB31</th>\n",
       "      <th>gb|JX423831.1|+|2676-3474|ARO:3001778|OXA-232</th>\n",
       "      <th>gb|AY536519.1|+|37-937|ARO:3002372|VEB-3</th>\n",
       "      <th>gb|KJ158441.1|+|2732-3377|ARO:3002789|QnrD2</th>\n",
       "      <th>gb|KF526113.1|+|0-1146|ARO:3002112|CMY-100</th>\n",
       "      <th>gb|X03988.1|+|382-1642|ARO:3000363|EreB</th>\n",
       "      <th>gb|AJ310778.1|+|7003-7573|ARO:3003015|dfrA19</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>genome_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>562.22444</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>562.42705</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>562.42733</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>562.42728</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>562.45644</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 159 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "res_gene   gb|U00096.3|-|3324062-3324911|ARO:3003386|Ecol_folP_SLF  \\\n",
       "genome_id                                                            \n",
       "562.22444                                                1.0         \n",
       "562.42705                                                1.0         \n",
       "562.42733                                                1.0         \n",
       "562.42728                                                1.0         \n",
       "562.45644                                                1.0         \n",
       "\n",
       "res_gene   gb|AP009048.1|+|3760295-3762710|ARO:3003303|Ecol_gyrB_AMU  \\\n",
       "genome_id                                                              \n",
       "562.22444                                                1.0           \n",
       "562.42705                                                1.0           \n",
       "562.42733                                                1.0           \n",
       "562.42728                                                1.0           \n",
       "562.45644                                                1.0           \n",
       "\n",
       "res_gene   gb|BA000007.3|+|4990267-4994296|ARO:3003288|Ecol_rpoB_RIF  \\\n",
       "genome_id                                                              \n",
       "562.22444                                                1.0           \n",
       "562.42705                                                1.0           \n",
       "562.42733                                                1.0           \n",
       "562.42728                                                1.0           \n",
       "562.45644                                                1.0           \n",
       "\n",
       "res_gene   gb|AH002539.2|+|1760-2945|ARO:3003368|Ecol_EFTu_KIR  \\\n",
       "genome_id                                                        \n",
       "562.22444                                                1.0     \n",
       "562.42705                                                0.0     \n",
       "562.42733                                                0.0     \n",
       "562.42728                                                1.0     \n",
       "562.45644                                                0.0     \n",
       "\n",
       "res_gene   gb|AE014075.1|-|3901532-3902762|ARO:3003369|Ecol_EFTu_PLV  \\\n",
       "genome_id                                                              \n",
       "562.22444                                                1.0           \n",
       "562.42705                                                0.0           \n",
       "562.42733                                                0.0           \n",
       "562.42728                                                1.0           \n",
       "562.45644                                                0.0           \n",
       "\n",
       "res_gene   gb|U00096.3|-|2336792-2339420|ARO:3003294|Ecol_gyrA_FLO  \\\n",
       "genome_id                                                            \n",
       "562.22444                                                1.0         \n",
       "562.42705                                                1.0         \n",
       "562.42733                                                1.0         \n",
       "562.42728                                                1.0         \n",
       "562.45644                                                1.0         \n",
       "\n",
       "res_gene   gb|AE006468.2|-|4038867-4041282|ARO:3003307|Sser_gyrB_FLO  \\\n",
       "genome_id                                                              \n",
       "562.22444                                                1.0           \n",
       "562.42705                                                0.0           \n",
       "562.42733                                                0.0           \n",
       "562.42728                                                1.0           \n",
       "562.45644                                                1.0           \n",
       "\n",
       "res_gene   gb|HG738867.1|+|2934266-2935658|ARO:3003890|Ecol_UhpT_FOF  \\\n",
       "genome_id                                                              \n",
       "562.22444                                                1.0           \n",
       "562.42705                                                1.0           \n",
       "562.42733                                                1.0           \n",
       "562.42728                                                1.0           \n",
       "562.45644                                                1.0           \n",
       "\n",
       "res_gene   gb|HG738867.1|-|2233288-2234647|ARO:3003889|Ecol_GlpT_FOF  \\\n",
       "genome_id                                                              \n",
       "562.22444                                                1.0           \n",
       "562.42705                                                1.0           \n",
       "562.42733                                                1.0           \n",
       "562.42728                                                1.0           \n",
       "562.45644                                                1.0           \n",
       "\n",
       "res_gene   gb|HG738867.1|+|2930707-2931298|ARO:3003893|Ecol_uhpA_FOF  ...  \\\n",
       "genome_id                                                             ...   \n",
       "562.22444                                                1.0          ...   \n",
       "562.42705                                                1.0          ...   \n",
       "562.42733                                                1.0          ...   \n",
       "562.42728                                                1.0          ...   \n",
       "562.45644                                                1.0          ...   \n",
       "\n",
       "res_gene   gb|X01702.1|+|484-1285|ARO:3002659|APH(6)-Ic  \\\n",
       "genome_id                                                 \n",
       "562.22444                                           0.0   \n",
       "562.42705                                           0.0   \n",
       "562.42733                                           0.0   \n",
       "562.42728                                           0.0   \n",
       "562.45644                                           0.0   \n",
       "\n",
       "res_gene   gb|JX259319.1|+|0-681|ARO:3002771|QnrB58  \\\n",
       "genome_id                                             \n",
       "562.22444                                       0.0   \n",
       "562.42705                                       0.0   \n",
       "562.42733                                       0.0   \n",
       "562.42728                                       0.0   \n",
       "562.45644                                       0.0   \n",
       "\n",
       "res_gene   gb|EU675686.2|+|3196-4042|ARO:3002616|aadA16  \\\n",
       "genome_id                                                 \n",
       "562.22444                                           0.0   \n",
       "562.42705                                           0.0   \n",
       "562.42733                                           0.0   \n",
       "562.42728                                           0.0   \n",
       "562.45644                                           0.0   \n",
       "\n",
       "res_gene   gb|HQ418999.2|+|0-681|ARO:3002746|QnrB31  \\\n",
       "genome_id                                             \n",
       "562.22444                                       0.0   \n",
       "562.42705                                       0.0   \n",
       "562.42733                                       0.0   \n",
       "562.42728                                       0.0   \n",
       "562.45644                                       0.0   \n",
       "\n",
       "res_gene   gb|JX423831.1|+|2676-3474|ARO:3001778|OXA-232  \\\n",
       "genome_id                                                  \n",
       "562.22444                                            0.0   \n",
       "562.42705                                            0.0   \n",
       "562.42733                                            0.0   \n",
       "562.42728                                            0.0   \n",
       "562.45644                                            0.0   \n",
       "\n",
       "res_gene   gb|AY536519.1|+|37-937|ARO:3002372|VEB-3  \\\n",
       "genome_id                                             \n",
       "562.22444                                       0.0   \n",
       "562.42705                                       0.0   \n",
       "562.42733                                       0.0   \n",
       "562.42728                                       0.0   \n",
       "562.45644                                       0.0   \n",
       "\n",
       "res_gene   gb|KJ158441.1|+|2732-3377|ARO:3002789|QnrD2  \\\n",
       "genome_id                                                \n",
       "562.22444                                          0.0   \n",
       "562.42705                                          0.0   \n",
       "562.42733                                          0.0   \n",
       "562.42728                                          0.0   \n",
       "562.45644                                          0.0   \n",
       "\n",
       "res_gene   gb|KF526113.1|+|0-1146|ARO:3002112|CMY-100  \\\n",
       "genome_id                                               \n",
       "562.22444                                         0.0   \n",
       "562.42705                                         0.0   \n",
       "562.42733                                         0.0   \n",
       "562.42728                                         0.0   \n",
       "562.45644                                         0.0   \n",
       "\n",
       "res_gene   gb|X03988.1|+|382-1642|ARO:3000363|EreB  \\\n",
       "genome_id                                            \n",
       "562.22444                                      0.0   \n",
       "562.42705                                      0.0   \n",
       "562.42733                                      0.0   \n",
       "562.42728                                      0.0   \n",
       "562.45644                                      0.0   \n",
       "\n",
       "res_gene   gb|AJ310778.1|+|7003-7573|ARO:3003015|dfrA19  \n",
       "genome_id                                                \n",
       "562.22444                                           0.0  \n",
       "562.42705                                           0.0  \n",
       "562.42733                                           0.0  \n",
       "562.42728                                           0.0  \n",
       "562.45644                                           0.0  \n",
       "\n",
       "[5 rows x 159 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print('\\nShape of PA df:', train_presence_absence_df.shape, '\\n')\n",
    "train_presence_absence_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c2856d94-8c99-400a-966f-694fb458ca99",
   "metadata": {},
   "source": [
    "### 2c. Also then need to subset our test data\n",
    "\n",
    "Always need to make sure our test data is in the same shape and format"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "f7d2dfc8-c5b6-4a52-9275-0cc60c829ee3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get just the unique 159 genes from our training data\n",
    "unique_train_genes = list(train_presence_absence_df.columns)\n",
    "test_presence_absence_df = build_gene_presence_absence(test_genes, y_test_ids, unique_train_genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "1b594936-8aeb-43b5-a2fc-84251a9dac00",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>res_gene</th>\n",
       "      <th>gb|U00096.3|-|3324062-3324911|ARO:3003386|Ecol_folP_SLF</th>\n",
       "      <th>gb|AP009048.1|+|3760295-3762710|ARO:3003303|Ecol_gyrB_AMU</th>\n",
       "      <th>gb|BA000007.3|+|4990267-4994296|ARO:3003288|Ecol_rpoB_RIF</th>\n",
       "      <th>gb|AH002539.2|+|1760-2945|ARO:3003368|Ecol_EFTu_KIR</th>\n",
       "      <th>gb|AE014075.1|-|3901532-3902762|ARO:3003369|Ecol_EFTu_PLV</th>\n",
       "      <th>gb|U00096.3|-|2336792-2339420|ARO:3003294|Ecol_gyrA_FLO</th>\n",
       "      <th>gb|AE006468.2|-|4038867-4041282|ARO:3003307|Sser_gyrB_FLO</th>\n",
       "      <th>gb|HG738867.1|+|2934266-2935658|ARO:3003890|Ecol_UhpT_FOF</th>\n",
       "      <th>gb|HG738867.1|-|2233288-2234647|ARO:3003889|Ecol_GlpT_FOF</th>\n",
       "      <th>gb|HG738867.1|+|2930707-2931298|ARO:3003893|Ecol_uhpA_FOF</th>\n",
       "      <th>...</th>\n",
       "      <th>gb|X01702.1|+|484-1285|ARO:3002659|APH(6)-Ic</th>\n",
       "      <th>gb|JX259319.1|+|0-681|ARO:3002771|QnrB58</th>\n",
       "      <th>gb|EU675686.2|+|3196-4042|ARO:3002616|aadA16</th>\n",
       "      <th>gb|HQ418999.2|+|0-681|ARO:3002746|QnrB31</th>\n",
       "      <th>gb|JX423831.1|+|2676-3474|ARO:3001778|OXA-232</th>\n",
       "      <th>gb|AY536519.1|+|37-937|ARO:3002372|VEB-3</th>\n",
       "      <th>gb|KJ158441.1|+|2732-3377|ARO:3002789|QnrD2</th>\n",
       "      <th>gb|KF526113.1|+|0-1146|ARO:3002112|CMY-100</th>\n",
       "      <th>gb|X03988.1|+|382-1642|ARO:3000363|EreB</th>\n",
       "      <th>gb|AJ310778.1|+|7003-7573|ARO:3003015|dfrA19</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>genome_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>562.42833</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>562.42739</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>562.22823</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>562.45646</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>562.22547</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 159 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "res_gene   gb|U00096.3|-|3324062-3324911|ARO:3003386|Ecol_folP_SLF  \\\n",
       "genome_id                                                            \n",
       "562.42833                                                1.0         \n",
       "562.42739                                                1.0         \n",
       "562.22823                                                1.0         \n",
       "562.45646                                                1.0         \n",
       "562.22547                                                1.0         \n",
       "\n",
       "res_gene   gb|AP009048.1|+|3760295-3762710|ARO:3003303|Ecol_gyrB_AMU  \\\n",
       "genome_id                                                              \n",
       "562.42833                                                1.0           \n",
       "562.42739                                                1.0           \n",
       "562.22823                                                1.0           \n",
       "562.45646                                                1.0           \n",
       "562.22547                                                1.0           \n",
       "\n",
       "res_gene   gb|BA000007.3|+|4990267-4994296|ARO:3003288|Ecol_rpoB_RIF  \\\n",
       "genome_id                                                              \n",
       "562.42833                                                1.0           \n",
       "562.42739                                                1.0           \n",
       "562.22823                                                1.0           \n",
       "562.45646                                                1.0           \n",
       "562.22547                                                1.0           \n",
       "\n",
       "res_gene   gb|AH002539.2|+|1760-2945|ARO:3003368|Ecol_EFTu_KIR  \\\n",
       "genome_id                                                        \n",
       "562.42833                                                0.0     \n",
       "562.42739                                                0.0     \n",
       "562.22823                                                1.0     \n",
       "562.45646                                                0.0     \n",
       "562.22547                                                1.0     \n",
       "\n",
       "res_gene   gb|AE014075.1|-|3901532-3902762|ARO:3003369|Ecol_EFTu_PLV  \\\n",
       "genome_id                                                              \n",
       "562.42833                                                0.0           \n",
       "562.42739                                                0.0           \n",
       "562.22823                                                1.0           \n",
       "562.45646                                                0.0           \n",
       "562.22547                                                1.0           \n",
       "\n",
       "res_gene   gb|U00096.3|-|2336792-2339420|ARO:3003294|Ecol_gyrA_FLO  \\\n",
       "genome_id                                                            \n",
       "562.42833                                                1.0         \n",
       "562.42739                                                1.0         \n",
       "562.22823                                                1.0         \n",
       "562.45646                                                1.0         \n",
       "562.22547                                                1.0         \n",
       "\n",
       "res_gene   gb|AE006468.2|-|4038867-4041282|ARO:3003307|Sser_gyrB_FLO  \\\n",
       "genome_id                                                              \n",
       "562.42833                                                0.0           \n",
       "562.42739                                                0.0           \n",
       "562.22823                                                1.0           \n",
       "562.45646                                                0.0           \n",
       "562.22547                                                0.0           \n",
       "\n",
       "res_gene   gb|HG738867.1|+|2934266-2935658|ARO:3003890|Ecol_UhpT_FOF  \\\n",
       "genome_id                                                              \n",
       "562.42833                                                1.0           \n",
       "562.42739                                                1.0           \n",
       "562.22823                                                1.0           \n",
       "562.45646                                                1.0           \n",
       "562.22547                                                1.0           \n",
       "\n",
       "res_gene   gb|HG738867.1|-|2233288-2234647|ARO:3003889|Ecol_GlpT_FOF  \\\n",
       "genome_id                                                              \n",
       "562.42833                                                1.0           \n",
       "562.42739                                                1.0           \n",
       "562.22823                                                1.0           \n",
       "562.45646                                                1.0           \n",
       "562.22547                                                1.0           \n",
       "\n",
       "res_gene   gb|HG738867.1|+|2930707-2931298|ARO:3003893|Ecol_uhpA_FOF  ...  \\\n",
       "genome_id                                                             ...   \n",
       "562.42833                                                1.0          ...   \n",
       "562.42739                                                1.0          ...   \n",
       "562.22823                                                1.0          ...   \n",
       "562.45646                                                1.0          ...   \n",
       "562.22547                                                1.0          ...   \n",
       "\n",
       "res_gene   gb|X01702.1|+|484-1285|ARO:3002659|APH(6)-Ic  \\\n",
       "genome_id                                                 \n",
       "562.42833                                             0   \n",
       "562.42739                                             0   \n",
       "562.22823                                             0   \n",
       "562.45646                                             0   \n",
       "562.22547                                             0   \n",
       "\n",
       "res_gene   gb|JX259319.1|+|0-681|ARO:3002771|QnrB58  \\\n",
       "genome_id                                             \n",
       "562.42833                                       0.0   \n",
       "562.42739                                       0.0   \n",
       "562.22823                                       0.0   \n",
       "562.45646                                       0.0   \n",
       "562.22547                                       0.0   \n",
       "\n",
       "res_gene   gb|EU675686.2|+|3196-4042|ARO:3002616|aadA16  \\\n",
       "genome_id                                                 \n",
       "562.42833                                             0   \n",
       "562.42739                                             0   \n",
       "562.22823                                             0   \n",
       "562.45646                                             0   \n",
       "562.22547                                             0   \n",
       "\n",
       "res_gene   gb|HQ418999.2|+|0-681|ARO:3002746|QnrB31  \\\n",
       "genome_id                                             \n",
       "562.42833                                       0.0   \n",
       "562.42739                                       0.0   \n",
       "562.22823                                       0.0   \n",
       "562.45646                                       0.0   \n",
       "562.22547                                       0.0   \n",
       "\n",
       "res_gene   gb|JX423831.1|+|2676-3474|ARO:3001778|OXA-232  \\\n",
       "genome_id                                                  \n",
       "562.42833                                              0   \n",
       "562.42739                                              0   \n",
       "562.22823                                              0   \n",
       "562.45646                                              0   \n",
       "562.22547                                              0   \n",
       "\n",
       "res_gene   gb|AY536519.1|+|37-937|ARO:3002372|VEB-3  \\\n",
       "genome_id                                             \n",
       "562.42833                                         0   \n",
       "562.42739                                         0   \n",
       "562.22823                                         0   \n",
       "562.45646                                         0   \n",
       "562.22547                                         0   \n",
       "\n",
       "res_gene   gb|KJ158441.1|+|2732-3377|ARO:3002789|QnrD2  \\\n",
       "genome_id                                                \n",
       "562.42833                                            0   \n",
       "562.42739                                            0   \n",
       "562.22823                                            0   \n",
       "562.45646                                            0   \n",
       "562.22547                                            0   \n",
       "\n",
       "res_gene   gb|KF526113.1|+|0-1146|ARO:3002112|CMY-100  \\\n",
       "genome_id                                               \n",
       "562.42833                                           0   \n",
       "562.42739                                           0   \n",
       "562.22823                                           0   \n",
       "562.45646                                           0   \n",
       "562.22547                                           0   \n",
       "\n",
       "res_gene   gb|X03988.1|+|382-1642|ARO:3000363|EreB  \\\n",
       "genome_id                                            \n",
       "562.42833                                        0   \n",
       "562.42739                                        0   \n",
       "562.22823                                        0   \n",
       "562.45646                                        0   \n",
       "562.22547                                        0   \n",
       "\n",
       "res_gene   gb|AJ310778.1|+|7003-7573|ARO:3003015|dfrA19  \n",
       "genome_id                                                \n",
       "562.42833                                           0.0  \n",
       "562.42739                                           0.0  \n",
       "562.22823                                           0.0  \n",
       "562.45646                                           0.0  \n",
       "562.22547                                           0.0  \n",
       "\n",
       "[5 rows x 159 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_presence_absence_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1d271974-bc3e-44c6-af50-9b3a2f94e086",
   "metadata": {},
   "source": [
    "## 3. Kmer Features\n",
    "\n",
    "Kmers are a representation of the raw sequencing data.\n",
    "\n",
    "K being a parameter setting the length of the sequence (e.g. 2-mer or 5-mer)\n",
    "\n",
    "There are two main options for kmerizing data:\n",
    "1. Utilize Python to count the sequence data\n",
    "2. Utilize a command line tool (see [Jellyfish](https://github.com/zippav/Jellyfish-2))\n",
    "\n",
    "In this tutorial we'll learn how to generate kmers manually on a small subset of data.\n",
    "\n",
    "Kmer counting is an expensive operation, for using Kmer features in your project - I've already generated an output file across all the data using Jellyfish (we'll loop back to this next week)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e2a9ca74-67e8-4150-bc50-f6983a1cc50c",
   "metadata": {},
   "source": [
    "### 3a. How can we count Kmers in a single sequence?"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ae0d71d5-261d-4514-86a9-30c22e84ad51",
   "metadata": {},
   "source": [
    "<div class=\"question\" style=\"color: #534646; background-color: #ffdfa3; padding: 1px; border-radius: 5px;\">\n",
    "\n",
    "#### Q. How would we generate the first 7-mer from a sequence?\n",
    "\n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4c10f9f4-67c6-491f-923f-6cdeef2edee1",
   "metadata": {},
   "source": [
    "We're going to need to break down our sequences into discrete chunks.\n",
    "\n",
    "As we saw in the presentation, we need to slide along our sequence doing this at each position:\n",
    "1. A simple for loop will work perfectly for this\n",
    "2. We can select each chunk of 7 nucleotides at each position\n",
    "3. We'll need to figure out how to keep track of this efficiently"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "05d68403-6b71-46a3-a7ba-c6d40d437c56",
   "metadata": {},
   "outputs": [],
   "source": [
    "def count_kmers(sequence, k=7):\n",
    "    \"\"\"\n",
    "    For a single sequence, slide over all nucleotides and count each k chunk\n",
    "\n",
    "    Args:\n",
    "      - sequence (str): raw nucleotide data\n",
    "      - k (int): size of kmers\n",
    "    \"\"\"\n",
    "    kmer_counts = defaultdict(int)\n",
    "    for i_start in range(len(sequence) - (k-1)):\n",
    "        kmer = sequence[i_start:i_start + k]\n",
    "        kmer_counts[kmer] += 1\n",
    "    \n",
    "    return kmer_counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "6f5054a1-8ecb-4077-9e0b-b6cdc169980e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "defaultdict(int,\n",
       "            {'ACGTGTG': 2,\n",
       "             'CGTGTGT': 1,\n",
       "             'GTGTGTA': 1,\n",
       "             'TGTGTAA': 1,\n",
       "             'GTGTAAG': 1,\n",
       "             'TGTAAGA': 1,\n",
       "             'GTAAGAC': 1,\n",
       "             'TAAGACG': 1,\n",
       "             'AAGACGT': 1,\n",
       "             'AGACGTG': 1,\n",
       "             'GACGTGT': 1,\n",
       "             'CGTGTGG': 1,\n",
       "             'GTGTGGC': 1,\n",
       "             'TGTGGCG': 1,\n",
       "             'GTGGCGA': 1})"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Test with a sequence\n",
    "test_sequence = 'ACGTGTGTAAGACGTGTGGCGA'\n",
    "count_kmers('ACGTGTGTAAGACGTGTGGCGA')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bb6c144e-657e-4024-a598-263bb894227e",
   "metadata": {},
   "source": [
    "### 3b. Apply this approach for each sample\n",
    "\n",
    "We have a function which can turn a sequence into counts of kmers\n",
    "\n",
    "This still isn't a usable \"feature\" for modeling though, we now need to apply across our training data, lets take a look at the data again:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "4ceff088-ed84-492d-a1b3-1ebdf7c0bac2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genome_id</th>\n",
       "      <th>contig</th>\n",
       "      <th>res_gene</th>\n",
       "      <th>match_start</th>\n",
       "      <th>match_end</th>\n",
       "      <th>match_qual</th>\n",
       "      <th>query_str</th>\n",
       "      <th>ref_gene_str</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>562.22444</td>\n",
       "      <td>562.22444.con.0012</td>\n",
       "      <td>gb|U00096.3|-|3324062-3324911|ARO:3003386|Ecol...</td>\n",
       "      <td>41274</td>\n",
       "      <td>42123</td>\n",
       "      <td>849M</td>\n",
       "      <td>TTACTCATAGCGTTTGTTTTCCTTTGCAGACAGAGTGGCTTCCACC...</td>\n",
       "      <td>TTACTCATAGCGTTTGTTTTCCTTTGCAGACAGAGTGGCTTCCACC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>562.22444</td>\n",
       "      <td>562.22444.con.0019</td>\n",
       "      <td>gb|AP009048.1|+|3760295-3762710|ARO:3003303|Ec...</td>\n",
       "      <td>60711</td>\n",
       "      <td>63126</td>\n",
       "      <td>2415M</td>\n",
       "      <td>ATGTCGAATTCTTATGACTCCTCCAGTATCAAAGTCCTGAAAGGGC...</td>\n",
       "      <td>ATGTCGAATTCTTATGACTCCTCCAGTATCAAAGTCCTGAAAGGGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>562.22444</td>\n",
       "      <td>562.22444.con.0030</td>\n",
       "      <td>gb|BA000007.3|+|4990267-4994296|ARO:3003288|Ec...</td>\n",
       "      <td>21837</td>\n",
       "      <td>25866</td>\n",
       "      <td>4029M</td>\n",
       "      <td>TTACTCGTCTTCCAGTTCGATGTTGATACCCAGCGAACGAATCTCT...</td>\n",
       "      <td>TTACTCGTCTTCCAGTTCGATGTTGATACCCAGCGAACGAATCTCT...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   genome_id              contig  \\\n",
       "0  562.22444  562.22444.con.0012   \n",
       "1  562.22444  562.22444.con.0019   \n",
       "2  562.22444  562.22444.con.0030   \n",
       "\n",
       "                                            res_gene  match_start  match_end  \\\n",
       "0  gb|U00096.3|-|3324062-3324911|ARO:3003386|Ecol...        41274      42123   \n",
       "1  gb|AP009048.1|+|3760295-3762710|ARO:3003303|Ec...        60711      63126   \n",
       "2  gb|BA000007.3|+|4990267-4994296|ARO:3003288|Ec...        21837      25866   \n",
       "\n",
       "  match_qual                                          query_str  \\\n",
       "0       849M  TTACTCATAGCGTTTGTTTTCCTTTGCAGACAGAGTGGCTTCCACC...   \n",
       "1      2415M  ATGTCGAATTCTTATGACTCCTCCAGTATCAAAGTCCTGAAAGGGC...   \n",
       "2      4029M  TTACTCGTCTTCCAGTTCGATGTTGATACCCAGCGAACGAATCTCT...   \n",
       "\n",
       "                                        ref_gene_str  \n",
       "0  TTACTCATAGCGTTTGTTTTCCTTTGCAGACAGAGTGGCTTCCACC...  \n",
       "1  ATGTCGAATTCTTATGACTCCTCCAGTATCAAAGTCCTGAAAGGGC...  \n",
       "2  TTACTCGTCTTCCAGTTCGATGTTGATACCCAGCGAACGAATCTCT...  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "train_genes.head(3)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4b0fa0cc-9148-4c93-ad4e-d9cbe8f9737f",
   "metadata": {},
   "source": [
    "<div class=\"question\" style=\"color: #534646; background-color: #ffdfa3; padding: 1px; border-radius: 5px;\">\n",
    "\n",
    "#### Q. What steps are we going to need to take from the data above?\n",
    "\n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "09607110-fbf4-49fe-bd65-045f179de571",
   "metadata": {},
   "source": [
    "Our data isn't in a single sequence per sample, it's per sample per gene.\n",
    "\n",
    "We can either:\n",
    "1. Count kmers per gene sequence and then aggregate per sample\n",
    "2. Aggregate sequences to a single sequence per sample then count\n",
    "\n",
    "We're going to take option 2 as it makes tracking the kmers much easier\n",
    "\n",
    "To make this computationally feasible we're going to try on just the first 5 samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "7e0dba0b-4f67-466e-b51a-72d04a36a92c",
   "metadata": {},
   "outputs": [],
   "source": [
    "def count_kmers_per_sample(dataset, sample_id_col='genome_id', gene_col='ref_gene_str'):\n",
    "\n",
    "    # Agg to full gene string per sample\n",
    "    seq_per_sample = dataset.groupby(sample_id_col).agg({'ref_gene_str': ''.join}).reset_index()\n",
    "    seq_per_sample['ref_gene_str'] = seq_per_sample['ref_gene_str'].str.upper()\n",
    "\n",
    "    kmer_counts = {}\n",
    "    for ref_name, sequence in seq_per_sample[[sample_id_col, gene_col]].to_records(index=False):\n",
    "        kmer_counts[ref_name] = count_kmers(sequence)\n",
    "\n",
    "    # Convert to DataFrame, fill empty with zero and transpose\n",
    "    kmer_counts_df = pd.DataFrame(kmer_counts).fillna(0).transpose()\n",
    "    \n",
    "    return kmer_counts_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "d988046d-e5f7-42dd-a645-731b2191707e",
   "metadata": {},
   "outputs": [],
   "source": [
    "first_5_samples = train_genes.genome_id.unique()[0:5]\n",
    "train_genes_first_five = train_genes[train_genes.genome_id.isin(first_5_samples)]\n",
    "test_kmer_counts = count_kmers_per_sample(train_genes_first_five)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "5a422cf7-9f62-485e-90ff-0ee98a51f131",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TTACTCA</th>\n",
       "      <th>TACTCAT</th>\n",
       "      <th>ACTCATA</th>\n",
       "      <th>CTCATAG</th>\n",
       "      <th>TCATAGC</th>\n",
       "      <th>CATAGCG</th>\n",
       "      <th>ATAGCGT</th>\n",
       "      <th>TAGCGTT</th>\n",
       "      <th>AGCGTTT</th>\n",
       "      <th>GCGTTTG</th>\n",
       "      <th>...</th>\n",
       "      <th>TTCTTAG</th>\n",
       "      <th>ATGCTAG</th>\n",
       "      <th>CTAGCAC</th>\n",
       "      <th>TAGCACT</th>\n",
       "      <th>TCTATGG</th>\n",
       "      <th>TTAGACC</th>\n",
       "      <th>GGACAAG</th>\n",
       "      <th>GACCCTA</th>\n",
       "      <th>TCCTATT</th>\n",
       "      <th>TTGGACC</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>562.22444</th>\n",
       "      <td>21.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>562.42705</th>\n",
       "      <td>19.0</td>\n",
       "      <td>200.0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>234.0</td>\n",
       "      <td>21.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>562.42728</th>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>26.0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>562.42733</th>\n",
       "      <td>10.0</td>\n",
       "      <td>199.0</td>\n",
       "      <td>47.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>224.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>562.45644</th>\n",
       "      <td>5.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>26.0</td>\n",
       "      <td>46.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 16245 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           TTACTCA  TACTCAT  ACTCATA  CTCATAG  TCATAGC  CATAGCG  ATAGCGT  \\\n",
       "562.22444     21.0      8.0      2.0      2.0      6.0      7.0     16.0   \n",
       "562.42705     19.0    200.0     18.0      3.0      8.0     10.0     15.0   \n",
       "562.42728      3.0      1.0      1.0      3.0      5.0      3.0     13.0   \n",
       "562.42733     10.0    199.0     47.0      5.0      9.0      6.0     15.0   \n",
       "562.45644      5.0      2.0      2.0      3.0      7.0     15.0     18.0   \n",
       "\n",
       "           TAGCGTT  AGCGTTT  GCGTTTG  ...  TTCTTAG  ATGCTAG  CTAGCAC  TAGCACT  \\\n",
       "562.22444      9.0     20.0     15.0  ...      0.0      0.0      0.0      0.0   \n",
       "562.42705      5.0    234.0     21.0  ...      0.0      0.0      0.0      0.0   \n",
       "562.42728     15.0     26.0     18.0  ...      0.0      0.0      0.0      0.0   \n",
       "562.42733     10.0    224.0     20.0  ...      0.0      0.0      0.0      0.0   \n",
       "562.45644     26.0     46.0     19.0  ...      1.0      1.0      1.0      1.0   \n",
       "\n",
       "           TCTATGG  TTAGACC  GGACAAG  GACCCTA  TCCTATT  TTGGACC  \n",
       "562.22444      0.0      0.0      0.0      0.0      0.0      0.0  \n",
       "562.42705      0.0      0.0      0.0      0.0      0.0      0.0  \n",
       "562.42728      0.0      0.0      0.0      0.0      0.0      0.0  \n",
       "562.42733      0.0      0.0      0.0      0.0      0.0      0.0  \n",
       "562.45644      1.0      1.0      1.0      1.0      1.0      1.0  \n",
       "\n",
       "[5 rows x 16245 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_kmer_counts.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8101451d-c123-4730-bce9-00653e7b1206",
   "metadata": {},
   "source": [
    "<div class=\"question\" style=\"color: #534646; background-color: #ffdfa3; padding: 1px; border-radius: 5px;\">\n",
    "\n",
    "#### Q. Any potential issues with the above?\n",
    "\n",
    "</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "6340c72f-2314-4163-bf66-57763c445878",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "16384"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Check unique possible 7-mers\n",
    "4**7"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "627c356f-8249-4aa2-9b3f-32594721d5ab",
   "metadata": {},
   "source": [
    "### Final Note\n",
    "\n",
    "We've counted every kmer that was present in each sample\n",
    "\n",
    "BUT: we haven't account for the kmers which weren't seen anywhere in the samples\n",
    "- 16,245 kmers present in the first 5 genomes\n",
    "- We need the feature matrix to be consistent for future unseen\n",
    "- Therefore need to account for ALL possible kmers\n",
    "\n",
    "Our feature matrix would need to be `N x 16384`\n",
    "\n",
    "This could get very large if we use a high K! (Remember 10-mers had over 1 million unique features)\n",
    "\n",
    "For next week I've already created a 11-mer matrix for both train and test using all unique genes seen across the data we have in this workshop."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3f2bdf53-896b-47be-9da4-fd4cbf5944cc",
   "metadata": {},
   "source": [
    "## 4. [BONUS] Sequence Features\n",
    "\n",
    "I've provided a simple method for featurizing the gene sequences themselves into a usable format. As mentioned it is an optional extra if you want to try using these features to build a Neural network as part of your project. It will be tricky to get it to train correctly and you may need to invest a lot more time to get it working (see the paper I linked in the slides). This featurization approach is likely insufficient alone but you can use it as a jumping off point if you wish to explore further.\n",
    "\n",
    "At a high level the process is:\n",
    "1. Extract only variant genes from CARD (ones which confer resistance through mutations)\n",
    "2. Subset genes to variant genes and randomly concatenate into a single long string\n",
    "3. Encode nucleotides to integers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "4a508387-955a-4321-a516-31a76617b263",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "nucleotide_fasta_protein_variant_model.fasta\n"
     ]
    }
   ],
   "source": [
    "# Load in just the varient genes and subset the data\n",
    "for file_name in os.listdir('../data/card_data/'):\n",
    "    if file_name.startswith('nucleotide_fasta_protein_variant_model'):\n",
    "        print(file_name)\n",
    "        with open(f'../data/card_data/{file_name}') as f:\n",
    "            fasta = f.readlines()\n",
    "variant_genes = [x.strip().split(' ')[0][1:] for x in fasta[0::2]]\n",
    "\n",
    "variant_gene_alignment_df = train_genes[train_genes.res_gene.isin(variant_genes)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "2ed2ff0d-ca91-4bed-9aa6-098965595aaa",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Precomputed maximum length of sequences\n",
    "max_length = 53859"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "9dea33e9-8345-44c1-8dbe-0a8e456a27e2",
   "metadata": {},
   "outputs": [],
   "source": [
    "def encode_seq(seq):\n",
    "    label_enc = {'A':1, 'C':2, 'G':3, 'T':4}\n",
    "    return [label_enc.get(x.upper(), 5) for x in seq]\n",
    "\n",
    "def featurize_variant_sequences(variant_genes, amr_max_length, pad_char=0):\n",
    "    gene_features = variant_genes.groupby('genome_id', sort=False)['ref_gene_str'].sum()\n",
    "    gene_features = [encode_seq(x) for x in gene_features]\n",
    "    gene_features = keras.utils.pad_sequences(gene_features, maxlen=max_length, padding='post', value=pad_char)\n",
    "       \n",
    "    return gene_features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "6a470145-a5e6-4b95-88c0-87750b7483b8",
   "metadata": {},
   "outputs": [],
   "source": [
    "sequence_features = featurize_variant_sequences(variant_gene_alignment_df, max_length)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "6fe7a2a4-61da-4ef2-8c2a-8aa09153d429",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[4, 4, 1, ..., 0, 0, 0],\n",
       "       [4, 4, 1, ..., 0, 0, 0],\n",
       "       [4, 4, 1, ..., 0, 0, 0],\n",
       "       [1, 4, 3, ..., 0, 0, 0],\n",
       "       [1, 4, 3, ..., 0, 0, 0]], dtype=int32)"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sequence_features[0:5]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5ce88661-ff0c-4862-87cf-fd25d77a76af",
   "metadata": {},
   "source": [
    "### Review:\n",
    "- This is a very simple featurization scheme to get started with\n",
    "- It randomly joins genes so the ordering is jumbled\n",
    "- We're only taking Variant genes, which may not be predictive for all samples\n",
    "  - Some may be mediated by the presence absence genes\n",
    "\n",
    "If planning to try using sequence features for the project:\n",
    "- Review the paper linked in the slides\n",
    "- Think about how to represent the data\n",
    "- Consider trying to build sequences using all genes if you have the computational resources to do so"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3d480a88-3838-45ae-ad9d-f02df8da63ef",
   "metadata": {},
   "source": [
    "## 5.Save data out for Assignment\n",
    "\n",
    "Save the presence/absence train/test data for use in the assignment (this is also available in the `train_test_data` course data folder if needed)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "48653ed4-3766-42ff-ab6e-0b48512f5326",
   "metadata": {},
   "outputs": [],
   "source": [
    "train_presence_absence_df.to_csv('../data/train_test_data/train_pa_genes.csv')\n",
    "test_presence_absence_df.to_csv('../data/train_test_data/test_pa_genes.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "38fba2cc-8049-458d-85b4-197c4783d6ca",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
